chpl/chpl-auth/src/main/java/gov/healthit/chpl/auth/controller/AuthenticationController.java

```diff
@@ -20,110 +24,160 @@
+        UpdatePasswordResponse response = new UpdatePasswordResponse();
+        if (Util.getCurrentUser() == null) {
+            throw new UserRetrievalException("No user is logged in.");
+        }
+
+        // get the current user
+        UserDTO currUser = userManager.getById(Util.getCurrentUser().getId());
+        if (currUser == null) {
+            throw new UserRetrievalException("The user with id " + Util.getCurrentUser().getId()
+                    + " could not be found or the logged in user does not have permission to modify their data.");
+        }
+
+        // check the strength of the new password
+        Strength strength = userManager.getPasswordStrength(currUser, request.getNewPassword());
+        if (strength.getScore() < UserManager.MIN_PASSWORD_STRENGTH) {
+            LOGGER.info("Strength results: [warning: {}] [suggestions: {}] [score: {}] [worst case crack time: {}]",
+                    strength.getFeedback().getWarning(),
+                    strength.getFeedback().getSuggestions().toString(),
+                    strength.getScore(),
+                    strength.getCrackTimesDisplay().getOfflineFastHashing1e10PerSecond());
+            response.setStrength(strength);
+            response.setPasswordUpdated(false);
+            return response;
+        }
+
+        // encode the old password passed in to compare
+        String currEncodedPassword = userManager.getEncodedPassword(currUser);
+        boolean oldPasswordMatches = bCryptPasswordEncoder.matches(request.getOldPassword(), currEncodedPassword);
+        if (!oldPasswordMatches) {
+            throw new UserRetrievalException("The provided old password does not match the database.");
+        } else {
+            userManager.updateUserPassword(currUser.getSubjectName(), request.getNewPassword());
+        }
+        response.setPasswordUpdated(true);
+        return response;
+    }
+
+    /**
+     * Reset a user's password and email them with the new one.
+     * @param userInfo the user whose password needs a reset
+     * @return "true" when password is reset
+     * @throws UserRetrievalException if user cannot be retrieved
+     * @throws MessagingException if email cannot be sent
+     */
+    @ApiOperation(value = "Reset a user's password.",
+            notes = "This service generates a new password, saves it to the user's account "
+                    + " and sends an email to the address associated with the user's account "
+                    + " containing the new password.")
+    @RequestMapping(value = "/reset_password", method = RequestMethod.POST,
+    consumes = MediaType.APPLICATION_JSON_VALUE,
+    produces = "application/json; charset=utf-8")
+    public String resetPassword(@RequestBody final UserResetPasswordJSONObject userInfo)
+            throws UserRetrievalException, MessagingException {
+
+        String newPassword = userManager.resetUserPassword(userInfo.getUserName(), userInfo.getEmail());
```

Comment:
Is this method resetting it to a password that would pass our strength test?

Reply:
I don't know. I didn't want to change current functionality. It probably is, as most of the 15 character random strings I generated to test things pass, but this could generate something with patterns or real words. I didn't think it was important to change it, because I'd expect a user to change their password when they get this email. I know I would; otherwise your actual password was transmitted in the clear.

That said, we probably should send them a link to create a new password instead of sending it in the clear; I think that's a no-no

Reply:
Ok so new ticket for that?

Reply:
I have it in my notes to bring up